### DISCUSSION TITLE
two ways to parse std::string to json

### DISCUSSION THREAD

**Discussion by 77h2l** [nlohmann/json#3258] [Q&A] [answered]
Hi,hope to get your reply soon again :)
I found that there are 2 ways to parse a string into json object; simple demo code below:
```cpp
std::string s;
json a(s);
```
---------
```
json b;
b.parse(s);
```

the first way barely costs time, but the later way costs a lot of time when the string was extremely log.
whats the difference between them?

**Comment by nlohmann** [comment]
The first snippet does not parse the string `s`, but just assigns it to a JSON value. So basically it is copied into the JSON value. Then you would call `a.dump()` you get the result of `s`, but with certain characters escape.

The second snippet does parse `s`, but the result is discarded - `parse` is a static function, so `b` is unchanged in your example.

See https://json.nlohmann.me/api/basic_json/parse/ for documentation and examples.

If you experience performance issues, please make sure you compile your code in Release mode. If there are still issues, please share the example so I can check.

**Reply by 77h2l** [reply]
you always reply so timely, awesome.
yeah I have found both two ways did not deserialization the input string.
actually my issue was as mentioned earlier  https://github.com/nlohmann/json/discussions/3202 
when I deserialization a 4k base64-encoded image string on arm64 device, it really costs. The same size of string on x86_64 server,it barely costs. 
I did compile my code in release mode,but the profile stills bottlenecks. actually I have tried several opensource json library like repid_json, simd json, yyjson and so on.yyjson performs better but your api was more friendly and easy to use. Hope this lib can improve performance,Thx :)

**Reply by nlohmann** [reply]
As I asked in #3202 - if you can share the original input, I can check where exactly there are bottlenecks. I also have to understand your usecase better, because maybe copying 4 KB into a DOM-like representation is not what you really need. However, if you already have successfully integrated RapidJSON or SIMD JSON into your code, then maybe working with them can quickly fix your issue.

**Reply by 77h2l** [reply]
the usecase was bellow.

**Reply by nlohmann** [reply]
What is the bottleneck for your ARM64 execution? Memory allocations?

**Comment by 77h2l** [comment, marked as answer]
[usecase.txt](https://github.com/nlohmann/json/files/7828386/usecase.txt)

**Reply by nlohmann** [reply]
Thanks - and what's the goal overall? Have everything in memory? Decode the image? I am asking because default parser is a general-purpose solution which does not bring the optimal performance for all usecases.

**Reply by 77h2l** [reply]
yap,actually, this kind of json string deserialization costs 200+ ms on the arm64 devices, and decode the image costs nearly 60 ms.
I just found that on the arm cpu,deserialization really costs.

**Reply by nlohmann** [reply]
How many ms with RapidJSON and how many with SIMD JSON?

**Reply by 77h2l** [reply]
actually I now use [yyjson](https://github.com/ibireme/yyjson) in my code, and on an equal footing, it costs 30 ms.

**Reply by 77h2l** [reply]
the strange thing is, I now use yyjson as the deserialization library, and use nlohmann json to format the output and dump into string, it was quite awkward.

**Comment by gregmarr** [comment]
json b;
    b.parse(s);

This spends a lot of time and gives you nothing in return, as `parse` is a static function that returns a `json` object.  You want to write it like this:

    auto b = json::parse(s);
